lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -290,6 +297,188 @@ defmodule Trento.Domain.SapSystem do
     }
   end
 
+  # Deregister an application instance and emit a ApplicationInstanceDeregistered
+  # also emit SapSystemDeregistered event if this was the last application instance
+  def execute(
+        %SapSystem{
+          sap_system_id: sap_system_id,
+          application: %Application{instances: instances}
+        } = sap_system,
+        %DeregisterApplicationInstance{
+          instance_number: instance_number,
+          sap_system_id: sap_system_id,
+          host_id: host_id,
+          deregistered_at: deregistered_at
+        } = command
+      ) do
+    instance = get_instance(instances, host_id, instance_number)
+
+    deregistered_instance_role =
+      case instance do
+        nil -> false
+        %Instance{features: features} -> get_application_instance_role(features)
+      end
+
+    sap_system
+    |> Multi.new()
+    |> Multi.execute(fn _ ->
+      %ApplicationInstanceDeregistered{
+        sap_system_id: sap_system_id,
+        instance_number: instance_number,
+        host_id: host_id,
+        deregistered_at: deregistered_at
+      }
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_application_instances_deregistered_events(
+        sap_system,
+        command,
+        deregistered_instance_role
+      )
+    end)
+    |> Multi.execute(fn sap_system ->
+      maybe_emit_sap_system_deregistered_event(sap_system, command)
```

Comment:
Do we send the whole command for some reason? Only the `deregistered_at` field is used in the command.
The same for the `maybe_` function above